nixos/modules/services/continuous-integration/buildkite-agents.nix

```diff
@@ -240,22 +255,18 @@ in
             KillMode = "mixed";
           };
       };
+  });
 
-    assertions = [
+  config.assertions = mapAgents (name: cfg: [
       { assertion = cfg.hooksPath == hooksDir || all (v: v == null) (attrValues cfg.hooks);
         message = ''
-          Options `services.buildkite-agent.hooksPath' and
-          `services.buildkite-agent.hooks.<name>' are mutually exclusive.
+          Options `services.buildkite-agents.${name}.hooksPath' and
+          `services.buildkite-agents.${name}.hooks.<name>' are mutually exclusive.
         '';
       }
-    ];
-  };
+  ]);
+
   imports = [
-    (mkRenamedOptionModule [ "services" "buildkite-agent" "token" ]                    [ "services" "buildkite-agent" "tokenPath" ])
-    (mkRenamedOptionModule [ "services" "buildkite-agent" "openssh" "privateKey" ]     [ "services" "buildkite-agent" "privateSshKeyPath" ])
-    (mkRenamedOptionModule [ "services" "buildkite-agent" "openssh" "privateKeyPath" ] [ "services" "buildkite-agent" "privateSshKeyPath" ])
-    (mkRemovedOptionModule [ "services" "buildkite-agent" "openssh" "publicKey" ]      "SSH public keys aren't necessary to clone private repos.")
-    (mkRemovedOptionModule [ "services" "buildkite-agent" "openssh" "publicKeyPath" ]  "SSH public keys aren't necessary to clone private repos.")
-    (mkRenamedOptionModule [ "services" "buildkite-agent" "meta-data"]                 [ "services" "buildkite-agent" "tags" ])
+    (mkRemovedOptionModule [ "services" "buildkite-agent"] "services.buildkite-agent has been moved to an attribute set at services.buildkite-agents")
```

Comment:
It should be relatively easy to make this backwards compatible with [`mkChangedOptionModule`](https://github.com/NixOS/nixpkgs/blob/faab2914f3c3ef888e0826ee7a58e2f55d120d22/lib/modules.nix#L727-L749), which will also issue a warning when the old option is used.

Reply:
We could do that, but what should we make the default agent name? The state paths for it would still change.

Reply:
Since people are expected to have used `dataDir` to refer to the path, I don't think it should be a problem. Also from this PR I can see that the service previously didn't even use `dataDir` to get its config, seemingly indicating that it was never fully supported anyways.

Also I correct myself: A simple
```nix
mkRenamedOptionModule [ "services" "buildkite-agent" ]
  [ "services" "buildkite-agents" "legacy" ]
```

should work (using the name `legacy`, or maybe `single` or `nixos-single` or so)

Reply:
the `dataDir` actually contains long-term state. People would have to move the contents of `/var/lib/buildkite-agent` to `/var/lib/buildkite-agent-legacy`, or get unexpected behavior. I think it's safer to have it break during the build, instead of during runtime.

Reply:
Ohh good point, didn't think of that. Okay so a `mkRenamedOptionModule` won't work, but a `mkChangedOptionModule` should still be possible with a `legacy` attribute (or similar) and by defaulting its `dataDir` to `/var/lib/buildkite-agent`

Reply:
I still think breaking old config and making sure the user did read the release notes is a better way forward here.

There have been quite some changes, and just silently assuming everything will still work as it did before is probably a bad idea anyways.

I also proposed to default to use `DynamicUsers`, and make the usage of a non-dynamic user, and custom `StateDirectory` the non-default case, which might cause nontrivial migrations aswell.

I think batching all these with the option syntax change is a reasonable path forward.

Reply:
What are the "quite some changes"? From what I can see all this PR does is to add support for multiple agents as it says, which imo really doesn't warrant breaking people's configs, especially if it only takes a couple lines to make it automatic.

Reply:
@Infinisil see the earlier PRs and the corresponding release notes:

Reply:
Were any of those not backwards compatible? If so they should've added a warning/error themselves. This PR itself seems 100% backwards compatible with a `mkChangedOptionModule`. Unless backwards compat *has* to be broken I'd rather not.

Reply:
Well, the module options themselves not, but the upgrade from buildkite 2 to 3 might have introduced some behavorial changes in how the bootstrap/hooks were configured: https://buildkite.com/docs/agent/v3/upgrading